docs/proposals/1964-pluggable-bbr-framework/README.md

```diff
@@ -0,0 +1,251 @@
+    GetPlugin(typeKey string) (bbrplugins.BBRPlugin, error)
+    GetFactories() map[string]PluginFactoryFunc
+    GetPlugins() map[string]bbrplugins.BBRPlugin
+    ListPlugins() []string
+    ListFactories() []string
+    CreatePlugin(typeKey string) (bbrplugins.BBRPlugin, error)
+    ContainsFactory(typeKey string) bool
+    ContainsPlugin(typeKey string) bool
+    String() string    //human readable string for logging
+}
+
+// PluginsChain is used to define a specific order of execution of the BBRPlugin instances stored in the registry
+// The BBRPlugin instances 
+type PluginsChain interface {
+    AddPlugin(typeKey string, registry PluginRegistry) error                    //to be added to the chain the plugin should be registered in the registry first
+    AddPluginAtInd(typeKey string, i int, r PluginRegistry) error               //only affects the instance of the plugin chain
+    GetPlugin(index int, registry PluginRegistry) (bbrplugins.BBRPlugin, error) //retrieves i-th plugin as defined in the chain from the registry
+    Length() int
+    ParseChatCompletion(data []byte) (openai.ChatCompletionNewParams, error)    //parses the bytes slice into an appropriate openai-go struct
+    ParseCompletion(data []byte) (openai.CompletionNewParams, error)            //likewise
+    GetSharedMemory(which string) interface{}  //returns an appropriate shared open-ai struct dependent on whether which 
+                                               //corresponds to Completion or ChatCompletion endpoint requested in the body 
+    Run(bodyBytes []byte, registry PluginRegistry) ([]byte, map[string]string, error) //return potentially mutated body and all headers map safely merged
+    String() string
+}
+//NOTE: for simplicity, in the initial PR, PluginsChain instance will be defined request only 
+```
+
+### Defaults
+
+```go
+
+const (
+    //A deafult plugin implementation of this plugin type will always be configured for request plugins chain
+    //Even though BBRPlugin type is not (yet) a K8s resource, it's logically akin to `kind`
+    //MUST start wit an upper case letter, use CamelNotation, only aplhanumericals after the first letter
+    PluginTypePattern   = `^[A-Z][A-Za-z0-9]*$`
+    MaxPluginTypeLength = 63
+    DefaultPluginType   = "MetaDataExtractor"
+    // Even though BBRPlugin is not a K8s resource yet, let's make its naming compliant with K8s resource naming
+    // Allows: lowercase letters, digits, hyphens, dots.
+    // Must start and end with a lowercase alphanumeric character.
+    // Middle characters group can contain lowercase alphanumerics, hyphens, and dots
+    // Middle and rightmost groups are optional
+    PluginNamePattern   = `^[a-z0-9]([-a-z0-9.]*[a-z0-9])?$`
+    DefaultPluginName   = "simple-model-extractor"
+    MaxPluginNameLength = 253
+    //Well-known custom header set to a model name
+    ModelHeader = "X-Gateway-Model-Name"
+)
+```
+
+### Current BBR reimplementation as BBRPlugin
```

Comment:
Not sure we need to have (most of) the code duplicated. Suffice to say the current implementation can be made compliant with the proposed interface.